clang/test/SemaSYCL/array-kernel-param.cpp

```diff
@@ -424,41 +424,19 @@ int main() {
-
-// Memcopy first array element
-// CHECK-NEXT: CallExpr
-// CHECK-NEXT: ImplicitCastExpr {{.*}} <BuiltinFnToFnPtr>
-// CHECK-NEXT: DeclRefExpr {{.*}} '<builtin fn type>' {{.*}} '__builtin_memcpy'
-// CHECK-NEXT: ImplicitCastExpr {{.*}} 'void *' <BitCast>
-// CHECK-NEXT: UnaryOperator {{.*}} 'StructWithSimplePointer *' prefix '&' cannot overflow
-// CHECK-NEXT: ArraySubscriptExpr {{.*}} 'StructWithSimplePointer':'StructWithSimplePointer' lvalue
-// CHECK-NEXT: ImplicitCastExpr {{.*}} 'StructWithSimplePointer *' <ArrayToPointerDecay>
-// CHECK-NEXT: MemberExpr {{.*}} 'StructWithSimplePointer[2]' lvalue .StructWithSimplePointerArray
-// CHECK-NEXT: DeclRefExpr {{.*}} '(lambda at {{.*}}array-kernel-param.cpp:126:9)' lvalue Var
-// CHECK-NEXT: IntegerLiteral {{.*}} 'unsigned long' 0
-// CHECK-NEXT: ImplicitCastExpr {{.*}} 'const void *' <BitCast>
+// CHECK-NEXT: CXXConstructExpr {{.*}} 'StructWithSimplePointer':'StructWithSimplePointer' 'void (const StructWithSimplePointer &) noexcept'
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'const StructWithSimplePointer' lvalue <NoOp>
+// CHECK-NEXT: UnaryOperator {{.*}} 'StructWithSimplePointer':'StructWithSimplePointer' lvalue prefix '*' cannot overflow
+// CHECK-NEXT: CXXReinterpretCastExpr {{.*}} 'StructWithSimplePointer *' reinterpret_cast<StructWithSimplePointer *> <BitCast>
 // CHECK-NEXT: UnaryOperator {{.*}} '__generated_StructWithSimplePointer *' prefix '&' cannot overflow
 // CHECK-NEXT: DeclRefExpr {{.*}} '__generated_StructWithSimplePointer' lvalue ParmVar {{.*}} '_arg_StructWithSimplePointerArray' '__generated_StructWithSimplePointer'
-// CHECK-NEXT: IntegerLiteral {{.*}} 'unsigned long' 16
-
-// Memcopy second array element
-// CHECK-NEXT: CallExpr
-// CHECK-NEXT: ImplicitCastExpr {{.*}} <BuiltinFnToFnPtr>
-// CHECK-NEXT: DeclRefExpr {{.*}} '<builtin fn type>' {{.*}} '__builtin_memcpy'
-// CHECK-NEXT: ImplicitCastExpr {{.*}} 'void *' <BitCast>
-// CHECK-NEXT: UnaryOperator {{.*}} 'StructWithSimplePointer *' prefix '&' cannot overflow
-// CHECK-NEXT: ArraySubscriptExpr {{.*}} 'StructWithSimplePointer':'StructWithSimplePointer' lvalue
-// CHECK-NEXT: ImplicitCastExpr {{.*}} 'StructWithSimplePointer *' <ArrayToPointerDecay>
-// CHECK-NEXT: MemberExpr {{.*}} 'StructWithSimplePointer[2]' lvalue .StructWithSimplePointerArray
-// CHECK-NEXT: DeclRefExpr {{.*}} '(lambda at {{.*}}array-kernel-param.cpp:126:9)' lvalue Var
-// CHECK-NEXT: IntegerLiteral {{.*}} 'unsigned long' 1
-// CHECK-NEXT: ImplicitCastExpr {{.*}} 'const void *' <BitCast>
+
+// CHECK: CXXConstructExpr {{.*}} 'StructWithSimplePointer':'StructWithSimplePointer' 'void (const StructWithSimplePointer &) noexcept'
+// CHECK-NEXT: ImplicitCastExpr {{.*}} 'const StructWithSimplePointer' lvalue <NoOp>
+// CHECK-NEXT: UnaryOperator {{.*}} 'StructWithSimplePointer':'StructWithSimplePointer' lvalue prefix '*' cannot overflow
+// CHECK-NEXT: CXXReinterpretCastExpr {{.*}} 'StructWithSimplePointer *' reinterpret_cast<StructWithSimplePointer *> <BitCast>
```

Comment:
When comparing AST here - https://godbolt.org/z/xYv9eEPbd, I see a `ParenExpr `being generated. Is that required?

Reply:
I don't think `ParenExpr` is necessary, especially for code that is generated by the compiler.
You can also remove parenthesis from normal code as well - https://godbolt.org/z/5Txsbx3Kr .

Reply:
OH! My bad. Lol I didn't notice the brackets in Godbolt example and for some reason I read `ParenExpr `as `ParentExpr` and was wondering what in the world that was. Thanks for clarifying!